builtin/game/chat.lua

```diff
@@ -855,25 +856,30 @@ core.register_chatcommand("status", {
 	end,
 })
 
+local function get_time(timeofday)
+	local time = math.floor(timeofday * 1440)
+	local minute = time % 60
+	local hour = (time - minute) / 60
+	return time, hour, minute
+end
+
 core.register_chatcommand("time", {
 	params = "[<0..23>:<0..59> | <0..24000>]",
 	description = "Show or set time of day",
 	privs = {},
 	func = function(name, param)
 		if param == "" then
-			local current_time = math.floor(core.get_timeofday() * 1440)
-			local minutes = current_time % 60
-			local hour = (current_time - minutes) / 60
-			return true, ("Current time is %d:%02d"):format(hour, minutes)
+			local _, current_hour, current_minute = get_time(core.get_timeofday())
+			return true, ("Current time is %d:%02d"):format(current_hour, current_minute)
 		end
 		local player_privs = core.get_player_privs(name)
 		if not player_privs.settime then
 			return false, "You don't have permission to run this command " ..
 				"(missing privilege: settime)."
 		end
-		local hour, minute = param:match("^(%d+):(%d+)$")
+		local relative, negative, hour, minute = param:match("^(~?)(-?)([%d~]+):([%d~]+)$")
```

Comment:
I'd prefer to see `hour < 0` checks rather than `negative == "-"` to simplify the regex as much as possible -> `([%d~-]+)`.